pkg/engine/lifecycle_test.go

```diff
@@ -982,3 +987,53 @@ func TestParallelRefresh(t *testing.T) {
 	assert.Equal(t, string(snap.Resources[3].URN.Name()), "resC")
 	assert.Equal(t, string(snap.Resources[4].URN.Name()), "resD")
 }
+
+func TestExternalRefresh(t *testing.T) {
+	loaders := []*deploytest.ProviderLoader{
+		deploytest.NewProviderLoader("pkgA", semver.MustParse("1.0.0"), func() (plugin.Provider, error) {
+			return &deploytest.Provider{}, nil
+		}),
+	}
+
+	// Our program reads a resource and exits.
+	program := deploytest.NewLanguageRuntime(func(_ plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+		_, _, err := monitor.ReadResource("pkgA:m:typA", "resA", "resA-some-id", "", resource.PropertyMap{}, "")
+		if !assert.NoError(t, err) {
+			t.FailNow()
+		}
+
+		return nil
+	})
+	host := deploytest.NewPluginHost(nil, program, loaders...)
+	p := &TestPlan{
+		Options: UpdateOptions{host: host},
+		Steps: []TestStep{
+			{
+				Op: Update,
+			},
+		},
+	}
+
+	// The read should place "resA" in the snapshot with the "External" bit set.
+	snap := p.Run(t, nil)
+	assert.Len(t, snap.Resources, 2)
+	assert.Equal(t, string(snap.Resources[0].URN.Name()), "default") // provider
+	assert.Equal(t, string(snap.Resources[1].URN.Name()), "resA")
+	assert.True(t, snap.Resources[1].External)
+
+	p = &TestPlan{
+		Options: UpdateOptions{host: host},
+		Steps: []TestStep{
+			{
+				Op: Refresh,
+			},
+		},
+	}
+
+	snap = p.Run(t, snap)
+	// A refresh should leave "resA" as it is in the snapshot. The External bit should still be set.
```

Comment:
might be better to save the new snapshot to a different variable and just assert that the resource lists are identical.

Reply:
the trouble is that all of the resource objects are pointers, so it doesn't make much sense to assert that they are equal since they already point to the same objects that are getting mutated.

Reply:
`assert.Equal` does not do pointer equality; it does deep equality: https://godoc.org/github.com/stretchr/testify/assert#Equalf

Reply:
Right - my point is that the first and second snapshots will always be trivially deep equal to one another because their contents are pointers to `resourceA`, which gets mutated by the second deployment. Unless we're doing some copying somewhere that I didn't know about.

Reply:
It's true that the number of elements may be different, but that's not the interesting part of this particular test.